lib/asciidoctor/helpers.rb

```diff
@@ -126,13 +126,23 @@ def self.uri_prefix str
   # Matches the characters in a URI to encode
   UriEncodeCharsRx = /[^\w\-.!~*';:@=+$,()\[\]]/
 
-  # Internal: Encode a String for inclusion in a URI.
-  #
-  # str - the String to URI encode
-  #
-  # Returns the String with all URI reserved characters encoded.
-  def self.uri_encode str
-    str.gsub(UriEncodeCharsRx) { $&.each_byte.map {|c| sprintf '%%%02X', c }.join }
+  # Internal: Encode a URI component String for safe inclusion in a URI.
+  #
+  # str - the URI component String to encode
+  #
+  # Returns the String with all reserved URI characters encoded (e.g., /, &, =, space, etc).
+  if RUBY_ENGINE == 'opal'
+    def self.encode_uri_component str
+      %x(
+        return encodeURIComponent(str).replace(/%20|[!'()*]/g, function (m) {
+          return m === '%20' ? '+' : '%' + m.charCodeAt(0).toString(16)
```

Comment:
At first I didn't understand why this was needed and then I read: https://unixpapa.com/js/querystring.html
I think we should explain why `encodeURIComponent` is not enough.

Reply:
Best would be to link to the following explanation, which is where I got this from: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent#Description (see last paragraph of section)